Topic/api 3/setup survey submit endpoint

Liriaene opened the pull request.

Liriaene committed:
API-3: Added survey file and added dependency

Added the survey.ts file that will handle the survey submission to the database.

Added a dependency to the package so we can run typescript from the command line without having to compile to javascript

Liriaene committed:
API-3: Added some logic for partial commit

Added some of the logic for a partial commit, however, running into some issues.

Liriaene committed:
API-3: Added survey post endpoint

Liriaene committed:
API-3: Updated more dependencies

These dependencies should allow the api to started with `npm run dev`

Liriaene committed:
API-3: Not sure if I should commit this

This commit has a bunch heavy handed "fixes" to allow the app.ts to compile to .js using the `run tsc dev` command while inside the API folder

Saposhiente committed:
API-3: Manualy cherry-pick

Cherry-picked the import fix and manually fixed the merge conflict

Liriaene committed:
API-3: Changed getting data to this ugly garbage

Having issuing getting the request data pulled in, so I switched to something that makes more sense to me for the time being (Granted it looks super awful). Needs to be chagned to something cleaner later

Liriaene committed:
API-3: Updated survey endpoitn

Liriaene committed:
API-3: hiding password to local db.

Bwamp. Remember to not upload passwords to remote kids.

The other stuff is test code

Liriaene committed:
API-3: Endpoint accepts JSON now

Endpoint accepts json now. No error handling yet. Also the sql calls need to be made dependent on eachother so their either all go in or don't.

ElijahRW committed:
Added minor npm build improvments

ElijahRW committed:
Replaced pg-promise with pg

ElijahRW committed:
Update to api documentation

ElijahRW committed:
Added Appointments Summary Query

ElijahRW committed:
Added appointment summary VIEW to DB

Liriaene committed:
API-3: Merged Dev into my branch

Liriaene committed:
API-3: SQL fix

Liriaene committed:
API-3: Completed survey sql, added newew dev toll

Finshed sql calls for survey submissions. Need to convet to transaction.
Added ts-node-dev package to allow api testing in live mod. script: `npm run dev`

Liriaene committed:
Merge remote-tracking branch 'origin/API/buildImprovements' into topic/API-3/SetupSurveySubmitEndpoint

Liriaene committed:
API-3: Changed survey endpoint to use async and transactions

This endpoint now uses transactions and async calls to submit to three tables.

Liriaene committed:
API-3: Renamed script to debug

renamed one of the dev scripts to debug.
Debug runs the server as typescript and auto respawns the server everytime a save is made to the source files.

ElijahRW commented on api/survey.ts:
```diff
@@ -0,0 +1,59 @@
+
+  // Take data from request and build sql call from it
+  // TODO: Move this data structure to it's own class
+  const {
```
this should be pushed out to a separate typescript file as a reference class.

gstone271 replied:
Instead see my comment on line 38

ElijahRW commented on api/app.ts:
```diff
@@ -19,6 +19,10 @@ app.use(
 )
 
+// POST endpoint for submiting surveys
+import surveyEndpoint from './survey'
```
End lines in ts with Semicolon
```suggestion
import surveyEndpoint from './survey';
```

ElijahRW commented on api/app.ts:
```diff
@@ -19,6 +19,10 @@ app.use(
 
+// POST endpoint for submiting surveys
+import surveyEndpoint from './survey'
+app.post('/appointment/survey', surveyEndpoint)
```
```suggestion
app.post('/appointment/survey', surveyEndpoint);
```

ElijahRW commented on api/survey.ts:
```diff
@@ -0,0 +1,59 @@
+export default handler(async (request : any) => {
+
+  let connection = db();
+  const client = await connection.connect()
```
I think you should use the connection object. It is already a usable for the queries.

ElijahRW commented on api/survey.ts:
```diff
@@ -0,0 +1,59 @@
+    throw error
+  } finally {
+    client.release()
+    return "Survey.ts finished."
```
Pull the return statement out of the finally clause. Putting it there _(if you comment out client.release())_ causes "Survey.ts finished.

Also, The return statement more communicative. (I.E. `Survey Successfully Submitted to DB`, or `Survey Submission Successful`, or `Survey {SURVEY_ID} has been successfully inserted`

ElijahRW commented on api/survey.ts:
```diff
@@ -0,0 +1,59 @@
+  const {
+    //Sharded Values
+    patientid,
+    completiondate,
```
I think these values should also be typed.

Liriaene replied:
Agreed, planned to move the definition out to it's own class first thorugh

gstone271 commented on DB Setup/createDB.sql:
```diff
@@ -97,3 +97,46 @@ create table Appointments (
 	PRIMARY KEY (AppointmentID)
 );
 
+CREATE OR REPLACE VIEW appointmentaummary AS 
```
misspelled

Liriaene replied:
I think I jacked up my branch commit history. Pretty sure this directory got deleted in a ticket that got merged already.

gstone271 commented on api/survey.ts:
```diff
@@ -0,0 +1,59 @@
+  } = request.body
+
+  var tfisurvey_sql = "INSERT INTO tfisurvey (tfisurveyid, patientid, completiondate, completiontime, tfi_i, tfi_sc, tfi_c, tfi_si, tfi_a, tfi_r, tfi_q, tfi_e, tfi_overallscore) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13)"
+  var tfisurvey_values = [tfisurveyid, patientid, completiondate, completiontime, tfi_i, tfi_sc, tfi_c, tfi_si, tfi_a, tfi_r, tfi_q, tfi_e, tfi_overallscore]
```
It would be nice to rewrite this function such that each of these strings (`tfi_i`, etc) only appears once per query it appears in. Eg:
`const tfiNames = ["tfisurveyid", "patientid", ...]`
`const tfisurveySql = 'INSERT INTO tfisurvey (' + tfiNames.join(', ') + ') VALUES (' + tfiNames.map((v, i) => '$' + (i + 1)).join(', ') + ')';`
`const tfisurveyValues = tfiNames.map((name) => request.body[name])`

gstone271 replied:
Bonus points if you do this in a loop or function so that you re-use this code for all of the survey tables, rather than copy/pasting.

gstone271 commented on api/survey.ts:
```diff
@@ -0,0 +1,59 @@
+  var thssurvey_sql = "INSERT INTO thssurvey (thssurveyid, patientid, completiondate, completiontime, ths_sectiona, ths_sectionb, ths_sectionc) VALUES ($1, $2, $3, $4, $5, $6, $7)"
+  var thssurvey_values = [thssurveyid, patientid, completiondate, completiontime, ths_sectiona, ths_sectionb, ths_sectionc]
+  
+  var tssurvey_sql = "INSERT INTO tssurvey (tssurveyid, patientid, completiondate, completiontime, ts_type) VALUES ($1, $2, $3, $4, $5)"
```
Always use `let` or `const` instead of `var`.

gstone271 commented on api/survey.ts:
```diff
@@ -0,0 +1,59 @@
+
+  try {
+    await client.query('BEGIN')
+    await connection.query(tfisurvey_sql, tfisurvey_values)
```
I think you can remove every `await` except the last one because you don't need to wait for one query to finish before sending the next one.

gstone271 replied:
Actually, use `await Promise.all(conn..., conn... ...)` to simultaneously wait for the three survey submit promises, rather than waiting for one at a time.

Liriaene committed:
API-3: Change requests

Liriaene committed:
API-3: Change request

Liriaene committed:
Merge branch 'development' into topic/API-3/SetupSurveySubmitEndpoint

Liriaene committed:
API-3: Changing endpoint to submitAppointment

After some discussion on Discord, we came to an agreement that this endpoint should really just be submitAppoint. Axing old file and changing app.ts to suppost this change

Liriaene committed:
API-3: Building out appointment sql

Building out appointment sql. Planning on moving the data structure out to somewhere else. I believe there will be a folder called "types" in the near future.

gstone271 commented on api/appointmentsPOST.ts:
```diff
@@ -0,0 +1,61 @@
+    await client.query('ROLLBACK')
+    throw error
+  } finally {
+    client.release()
```
Is this going to disconnect the connection and prevent its reuse?

Liriaene replied:
To my understanding no. This release releases a client instance created earlier from the shared pool.

gstone271 commented on api/appointmentsPOST.ts:
```diff
@@ -0,0 +1,61 @@
+
+  try {
+    await client.query('BEGIN')
+    connection.query(tfisurvey_sql, tfisurvey_values)
```
You seem to use `client` and `connection` interchangeably. I suspect you can just eliminate `client`.

gstone271 replied:
will be resolved by updating to withConnection

Liriaene committed:
Merge branch 'development' into topic/API-3/SetupSurveySubmitEndpoint

Liriaene committed:
API-3: Fixed appointment submit name. Added appointment insert

Fixed the appointment submit endpoint.
Added sql and data fields for appointments table. Having issues with getting DB to comply though

Liriaene committed:
API-3: Updated post enpoint also submit audiologistexams

Endpoint now sends data to 5 tables.
Special Thanks for Elijah for the assist.

Liriaene committed:
API-3: Files shouldn't be here

I probably jacked up my commit tree into a monstrosity. These files were already deleted in a different ticket.

Liriaene committed:
API-3: Changed logic to now use promise instead of multipul await calls

wynnliam commented on api/db.ts:
```diff
@@ -1,4 +1,16 @@
+  user: 'postgres',
+  host: 'localhost',
+  database: 'va_trial',
+  password: 'postgres',
```
Is there some way to hide the password when doing a commit? I know we'll be using a config file or something in the future. I suppose you can just comment out the password before pushing commits?

Anyways, everything else looks good to me, and I don't think having a dummy password should prevent a merge.

gstone271 replied:
Good thinking; my authenticate branch (which was prematurely merged, you should review it just to see how it works) solves this problem

gstone271 commented on api/appointmentsPOST.ts:
```diff
@@ -0,0 +1,86 @@
+
+export default handler(async (request: any) => {
+
+  let connection = db();
```
This needs to be updated to `withConnection` form; rebase onto development so that you get `withConnection` then fix this.

gstone271 commented on api/appointmentsPOST.ts:
```diff
@@ -0,0 +1,86 @@
+
+      return { tfisurveyid,  thssurveyid, tssurveyid};
+    })
+  } catch (error) {
```
This `catch` is redundant; just don't catch the error, and it will simply be thrown.

gstone271 commented on api/appointmentsPOST.ts:
```diff
@@ -0,0 +1,86 @@
+  } catch (error) {
+    throw error;
+  } finally {
+    client.release();
```
This will become unnecessary once you update to `withConnection`.

gstone271 commented on api/appointmentsPOST.ts:
```diff
@@ -0,0 +1,86 @@
+
+  // Appointments inserted last because it needs info from the other tables
+  try {
+    await Promise.all([
```
Instead of using `.then`, say `const values = await Promise.all(...)` followed by the line 69 stuff and so forth.

gstone271 commented on api/appointmentsPOST.ts:
```diff
@@ -0,0 +1,86 @@
+      let appointment_values = [authorityid, patientid, tfisurveyid, thssurveyid, tssurveyid, audiologistexamsid]
+      client.query(appointment_sql, appointment_values)
+
+      return { tfisurveyid,  thssurveyid, tssurveyid};
```
This `return` will not return from this `handler` function; instead it only sets the return value of the Promise on line 63, but this value is then discarded because the Promise was not sent anywhere. (Or maybe it's sent anyway by an implicit return? But we don't want to use an implicit return.) Either way, applying my change for line 63 will fix this.

Saposhiente committed:
Shorted sql calls down

Liriaene committed:
API-3: Condensed sql to use more funtional style programming

gstone271 commented:
Redundant with #28
